pages/glossary/keyboard-shortcut.md

```diff
@@ -0,0 +1,17 @@
+---
+title: Keyboard shortcut
+key: keyboard-shortcut
+unambiguous: true
+objective: true
+---
+
+A _keyboard shortcut_ is a sequence of [events][event] of the type [keydown][] and [keyup][], that:
+ - begins with a [keydown][] [event][]; and
+ - ends when the number of [keyup][] [events][event] is equal to the number of [keydown][] [events][event]; and
+ - are [dispatched][] to an [event target][].
```

Comment:
I like this but… need more work…

---
The balanced sequence of ups and downs may need some sort of maximality condition. Otherwise:
* in (keydown "a")(keydown "b")(keyup "b")(keyup "a"), there are two shortcuts:
  + (keydown "a")(keydown "b")(keyup "b")(keyup "a")
  + (keydown "b")(keyup "b")

  Which is mildly annoying but OK-ish.
* in (keydown "a")(keydown "b")(keyup "a")(keyup "b"), there are two shortcuts:
  + (keydown "a")(keydown "b")(keyup "b")(keyup "b")
  + (keydown "b")(keyup "a")

  Which is very annoying.

And the maximality condition must be fine-tuned enough so that (keydown "a")(keyup "a")(keydown "b")(keyup "b") is not merged into a single shortcut.

---
The examples in the rules are actually reacting to the keydown events: https://github.com/act-rules/act-rules.github.io/blob/c0f713674868ae75bb9c53372d3e6108b5577992/test-assets/ffbc54/shortcut.js#L12
So the keyups are not really part of the shortcut. Note that the events could instead react to the keyup events and  still work, so it's not about changing everything to only keydown events.

This is somewhat in contradiction with the maximality needed in the first point…

---
I feel there are two things mixed in this definition, making it not match the mundane idea of shorcut:
* one is the sequence of events (the cat walking on the keyboard, essentially);
* the second is the page reacting to these events.

I feel that "shortcut" applies to what the page does. Me pushing keys on my board is me activating that shortcut, but it's not the shortcut proper.

And this may actually help mitigating both previous points…

I think one way to improve that could be to split the definition:
* a "keyboard activation/interaction/whatever" would be the sequence of events. On that, the maximality condition is needed (so that it does have a clear end and can't be "what if another key is going to be pressed and released later?") The maximality condition also make sense because as long it's not met it means I'm stilling keeping keys pressed, and it is fair to say I thus haven't finished my interaction with the keyboard.
* a "keyboard shortcut" which is the page reacting to a keyboard interaction. That one is tricky because the shortcut is likely going to respond to prefixes of the interaction and because it can react to various interactions. By listening to keydown/up events, the reaction happens on a single event, not on the full sequence (the listener is triggered by individual event) and the effect is usually decided on a state, not on what lead to it (we look whether "Ctrl" is still pressed when we receive the event, but we don't care about what happened previously).
  + This does prevent the shortcut to react to ful interactions only. For example, in (keydown "Ctrl")(keydown "a")(keyup "Ctrl")(keyup "a") if reacting only to (keyup "a"), then the "Ctrl" modifier is not present when reacting.
  + Another weird case which works with the examples in the rules and need to be handled correctly is (keydown "b")(keydown "Ctrl")(keydown "shift")(keyup "b")(keydown "c")(keyup "c")(keydown "a") REACTION IS HERE (keyup "shift")(keydown "d")(keyup "Ctrl")(keyup "a")(keydown "Alt")(keyup "Alt")(keyup "d") BALANCED SEQUENCE ENDS HERE
Note that in this interaction, "shift" is still pressed when reaction occurs, as the event handler doesn't care.

So, there's still a lot of difficulty to solve in the definition of keyboard shortcut 😐